Add unit tests for pip commands not using cwd

deveshks opened the pull request:
Follow-up of https://github.com/pypa/pip/pull/7955 . Fixes and closes #7731, closes https://github.com/pypa/pip/issues/2926, closes https://github.com/pypa/pip/issues/3710 and closes https://github.com/pypa/pip/issues/7971

Add unit tests for check, freeze, show and install to verify that  the behaviour of `python -m pip` is same as `pip` when run from the current working directory, and verifying that adding current working directory to PYTHONPATH displays the appropriate behaviour with `python -m pip`

deveshks commented:
Hi @sbidoul , @xavfernandez 

This PR is a followup of https://github.com/pypa/pip/pull/7955 to add unit tests for the remaining 4 commands affected by this on the same principle. 

Would really appreciate if you can take a look at the same and share your thoughts 😊

deveshks commented on tests/functional/test_check.py:
```diff
@@ -239,3 +239,56 @@ def test_basic_check_broken_metadata(script):
+
+    # Check should not complain about broken requirements
+    # when run from package directory
+    result = script.pip('check')
```
Added `cwd=pkg_path` to run from `pkg_path`

deveshks commented on tests/functional/test_check.py:
```diff
@@ -239,3 +239,56 @@ def test_basic_check_broken_metadata(script):
+
+    # Check should mention about missing requirement simple
+    # when run from package directory
+    result = script.pip('check', expect_error=True)
```
Added `cwd=pkg_path` to run from `pkg_path`

deveshks commented:
Hi @xavfernandez 

I have addressed the review comments. Please let me know if there are further changes needed.

sbidoul commented on tests/functional/test_check.py:
```diff
@@ -239,3 +239,56 @@ def test_basic_check_broken_metadata(script):
+    script.run('python', 'setup.py', 'egg_info',
+               expect_stderr=True, cwd=pkg_path)
+
+    # Add PYTHONPATH env variable
```
This comment does not add much value ;)

deveshks replied:
Removed

sbidoul commented on tests/functional/test_check.py:
```diff
@@ -239,3 +239,56 @@ def test_basic_check_broken_metadata(script):
+    script.environ.update({'PYTHONPATH': pkg_path})
+
+    # Check should mention about missing requirement simple
+    # when run from package directory
```
```suggestion
    # when run from package directory, when package directory is in PYTHONPATH
```

deveshks replied:
Updated the comment

sbidoul commented on tests/functional/test_freeze.py:
```diff
@@ -818,9 +819,38 @@ def test_freeze_path_multiple(tmpdir, script, data):
-    assert req.startswith("simple @ file://")
-    script.pip("install", req)
-    result = script.pip("freeze")
-    assert req in result.stdout
```
Why removing this test?

deveshks replied:
Aah good catch, I removed it by mistake. Readding it

sbidoul commented on tests/functional/test_freeze.py:
```diff
@@ -818,9 +819,38 @@ def test_freeze_path_multiple(tmpdir, script, data):
+
+    # Freeze should not include package simple when run from package directory
+    result = script.pip('freeze', cwd=pkg_path)
+    assert 'simple==1.0' not in result.stdout
```
```suggestion
    assert 'simple==' not in result.stdout
```

deveshks replied:
Fixed

sbidoul commented on tests/functional/test_freeze.py:
```diff
@@ -818,9 +819,38 @@ def test_freeze_path_multiple(tmpdir, script, data):
+    # Add PYTHONPATH env variable
+    script.environ.update({'PYTHONPATH': pkg_path})
+
+    # Freeze should include package simple when run from package directory
```
Similar remark as above.

deveshks replied:
Updated the comment

sbidoul commented on tests/functional/test_install.py:
```diff
@@ -1821,3 +1821,51 @@ def test_install_sends_client_cert(install_args, script, cert_factory, data):
+    result = script.pip('install', '--find-links',
+                        data.find_links, 'simple',
+                        expect_stderr=True, cwd=pkg_path)
+    assert 'Successfully installed simple' in result.stdout
```
Check that `Requirement already satisfied: simple` is not in result?

deveshks replied:
Added the check

sbidoul commented on tests/functional/test_install.py:
```diff
@@ -1821,3 +1821,51 @@ def test_install_sends_client_cert(install_args, script, cert_factory, data):
+    # Add PYTHONPATH env variable
+    script.environ.update({'PYTHONPATH': pkg_path})
+
+    # Uninstalling the package and installing it again will fail
```
Similar remark as above regarding the comments.

deveshks replied:
Updated the comment

deveshks commented:
Hi @sbidoul 

I have addressed the review comments. Please let me know if there are further changes needed.

deveshks commented on tests/functional/test_install.py:
```diff
@@ -1821,3 +1821,53 @@ def test_install_sends_client_cert(install_args, script, cert_factory, data):
+    script.pip('install', '-e', '.',
+               expect_stderr=True, cwd=pkg_path)
+
+    # Uninstall will fail with given warning
```
I have reworded the comment accordingly. Please take a look

xavfernandez replied:
This comment still seems misleading to me :thinking: 
The `script.pip('uninstall', 'simple', '-y')` does not seem to fail or print warnings since there is no `expect_stderr=True`

deveshks replied:
Created https://github.com/pypa/pip/pull/8012 to address this comment.

deveshks commented on tests/functional/test_install.py:
```diff
@@ -1821,3 +1821,53 @@ def test_install_sends_client_cert(install_args, script, cert_factory, data):
+
+    script.environ.update({'PYTHONPATH': pkg_path})
+
+    # Uninstalling the package and installing it again will fail,
```
I have reworded the comment accordingly. Please take a look

xavfernandez replied:
There is no uninstalling here

deveshks replied:
Created https://github.com/pypa/pip/pull/8012 to address this comment.

deveshks commented:
> @deveshks thanks for your patience. I realize there were a couple comments I had not submitted, sorry about that.

No worries @sbidoul . I have addressed the review comments in my latest commit. Please take a look.

deveshks commented:
Thanks @sbidoul  for the approval.

Hi @xavfernandez , @pradyunsg ,

If the changes look good, could I please get this PR merged as well ?

sbidoul commented:
@deveshks there were two additional comments by @xavfernandez after I approved. Can you attend them in a separate PR?

deveshks commented:
> @deveshks there were two additional comments by @xavfernandez after I approved. Can you attend them in a separate PR?

Hi @sbidoul , @xavfernandez 

Thanks for catching that. It seems like the commit I made to address rewording the comments didn't show up here. As suggested, I have created a new PR to address the comments at https://github.com/pypa/pip/pull/8012 . Please take a look 😊
